### PR TITLE
Issue 118

### PR DESCRIPTION
**Pull request by VCP-Development**
Allow for uploading of an image file (iso, qcow2) as part of the vnfd tar file.  There is an associated change made in the openstack4j-plugin repository.

### CODE REVIEW

File: api/src/main/java/org/openbaton/nfvo/api/catalogue/RestVNFPackage.java [hunk 1]
```diff
@@ -87,7 +93,136 @@ public String onboard(
...
+      File tarFile = new File(directory + name);
+      file.transferTo(tarFile);
+      log.debug("File size is " + file.getSize());
+
+      Runtime rt = Runtime.getRuntime();
+
+      try {
+        // untar the file
+        Process p1 =
+            rt.exec(new String[] {"bash", "-c", "tar -xf " + name}, new String[0], tempDirectory);
```
Comment (gc4rella) [inline api/src/main/java/org/openbaton/nfvo/api/catalogue/RestVNFPackage.java]: @RuthDVerizon I'm a bit worried about using bash commands directly in the NFVO. Main reason is that there are some users using windows, and this particular command will probably fail and throw some exceptions. 
In addition, I see that some parts of this code is doing more or less the same which is done in the VNFPackageManagement ( https://github.com/openbaton/NFVO/blob/8e7206efa6a72ad315725062d4bb9019e64e9afe/core/core-impl/src/main/java/org/openbaton/nfvo/core/api/VNFPackageManagement.java#L105) class. In that case, we are making use of some java libraries for extracting content from the VNF Package, without requiring the execution of bash commands.

**Comment by RuthDVerizon** [comment]
@gc4rella The reason that the decision was made to extract the file before loading it into the program was that when we loaded sizeable files the program ran out of stack space.  By using the file system to do the work we avoided that problem.  

Can I ask what operating systems you plan on supporting?

**Comment by gc4rella** [comment]
@RuthDVerizon I see the point...but probably we need to think a bit more about a slightly different approach which could work out of the box on different operating systems. 

we plan to support at least OS X and Windows. For OS X OB runs on it without any modifications. I think @marcellom was doing some tests and encountered some issue after your changes. For Windows, there are some paths which need to be updated, but with very minimal effort (https://github.com/openbaton/NFVO/issues/68) some of our users were able to execute OB without any issues. 

All in all, we will need to update probably also the bootstrap script (https://github.com/openbaton/bootstrap), so that additional components (apache) and additional folders are created in advance. We can support on that, once we agree on the right approach for moving forward.

**Comment by gc4rella** [comment]
....or keep everything as it is but provide this version of the plugin as an alternative...this should also be easily doable starting the plugin with a different name and call this particular version from the pop.json file using "name.type". Would you like this solution?

**Comment by RuthDVerizon** [comment]
I uploaded a new file api/src/main/java/org/openbaton/nfvo/api/catalogue/RestVNFPackage.java which has some significant changes to how some of the issues with the special tar commands are executed.  I also changed it so that the paths are configurable so that it can accommodate any location including in windows (at least in theory, since I do not have a windows box to test on I cannot confirm). 

There is still one issue that I have not been able to solve.  The processing of the initial tar file.  If we load the file into the program we run a risk of running out of stack space depending on the size of the file.  If we are using only Linux based systems it should not be a problem to use the linux files, the issue would be Windows.  We could attempt to use system commands and then load if that fails.  That should work for small images but not larger images.  Do you have any ideas for how we might solve this issue?

**Comment by RuthDVerizon** [comment]
@gc4rella That is certainly an option but I don't believe it is necessary.  The code here will attempt to untar the file and see if there is an image inside.  If that fails it will process the tar file exactly as it used to.

**Comment by RuthDevlaeminck** [comment]
Per discussion held at Verizon a different approach will be used to upload images, closing this pull request.
--

**Pull request closed without merging**